### PR TITLE
Split fileToUpload struct into separate file and dir structs

### PR DESCRIPTION
**Pull request by bduffany**
While reading this code, I found the fileToUpload struct to be confusing because it can represent either a file or a directory. 

This PR adds a separate dirToUpload struct and simplifies the code by removing the conditional logic checking whether a fileToUpload is a file or directory.

**Related issues**: N/A

### CODE REVIEW

File: enterprise/server/remote_execution/dirtools/dirtools.go [hunk 1]
```diff
@@ -219,92 +216,85 @@ func newDirToUpload(instanceName string, digestFunction repb.DigestFunction_Valu
...
 		// Note: uploader.Upload closes the file after it is uploaded.
-		if err := uploader.Upload(uploadableFile.resourceName.GetDigest(), rsc); err != nil {
+		if err := uploader.Upload(uploadableFile.digest, f); err != nil {
 			return err
 		}
 	}
+	return nil
+}
 
+func uploadDirs(ctx context.Context, uploader *cachetools.BatchCASUploader, dirsToUpload []*dirToUpload) error {
```
Comment (sluongng) [inline enterprise/server/remote_execution/dirtools/dirtools.go:291]: Nit: I think moving this inside UploadTree would make the whole thing easier to follow vs using a separate func.

**Review by sluongng — APPROVED** [review]
LGTM.

I also started doing this initially. Then after a bit more refactoring, I realized that it’s easier to follow the logic after I in-lined `DirectoryNode()` and got rid of the 2 structs `fileToUpload` and `dirToUpload` completely. Hope that future PRs will head toward the same direction.

File: enterprise/server/remote_execution/dirtools/dirtools.go (hunk 1 above)
Comment (bduffany) [reply enterprise/server/remote_execution/dirtools/dirtools.go]: done

**Comment by bduffany** [comment]
> Then after a bit more refactoring, I realized that it’s easier to follow the logic after I in-lined DirectoryNode() and got rid of the 2 structs fileToUpload and dirToUpload completely. Hope that future PRs will head toward the same direction.

Yeah, I would like to eventually reach this state, but would like to gradually transform the code in ways where each change is more digestable and then once the code is in a state we're happy with, make the big 2 changes where we stop uploading directories unnecessarily and also make sure we track all uploads and/or call FindMissing before doing uploads.

**Review by iain-macdonald — APPROVED** [review]
LGTM, Approval

**Pull request merged**